gradoop-data-integration/src/main/java/org/gradoop/dataintegration/transformation/VertexDeduplication.java

```diff
@@ -0,0 +1,113 @@
+import org.gradoop.flink.model.impl.functions.epgm.ByLabel;
+import org.gradoop.flink.model.impl.functions.epgm.SourceId;
+import org.gradoop.flink.model.impl.functions.epgm.TargetId;
+import org.gradoop.flink.model.impl.functions.filters.Not;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Deduplicates vertices based on some attribute.<p>
+ * Given a label and a list of property keys, this operator will find all vertices of that label and condense
+ * duplicates to one vertex. Two (or more) vertices are considered duplicates if all property values,
+ * respective to the list of given keys, are equal.<p>
+ * Other attributes of the vertices are ignored, the new (condensed) vertex will have the attributes of some
+ * (random) vertex of its duplicates.
+ * <p>
+ * This will create no new elements and retain the graph head.
+ *
+ * @param <G>  The graph head type.
+ * @param <V>  The vertex type.
+ * @param <E>  The edge type.
+ * @param <LG> The type of the graph.
+ * @param <GC> The type of the graph collection.
+ */
+public class VertexDeduplication<
+  G extends GraphHead,
+  V extends Vertex,
+  E extends Edge,
+  LG extends BaseGraph<G, V, E, LG, GC>,
+  GC extends BaseGraphCollection<G, V, E, LG, GC>>
+  implements UnaryBaseGraphToBaseGraphOperator<LG> {
+
+  /**
+   * The label of vertices to deduplicate.
+   */
+  private final String label;
+
+  /**
+   * The list of property keys used to determine equal vertices.
+   */
+  private final List<String> propertyKeys;
+
+  /**
+   * Creates a new instance of this vertex deduplication operator.
+   *
+   * @param label        The label of vertices to deduplicate.
+   * @param propertyKeys The properties used to check if two (or more) vertices of that label are duplicates.
+   */
+  public VertexDeduplication(String label, List<String> propertyKeys) {
+    this.label = Objects.requireNonNull(label);
+    this.propertyKeys = Objects.requireNonNull(propertyKeys);
+  }
+
+  @Override
+  public LG execute(LG graph) {
+    DataSet<V> annotatedVertices = graph.getVerticesByLabel(label)
+      .groupBy(new GetPropertiesAsList<>(propertyKeys))
+      .reduceGroup(new MarkDuplicatesInGroup<>())
+      .name("Mark duplicate vertices");
+    DataSet<Tuple2<GradoopId, GradoopId>> vertexToDedupVertex = annotatedVertices
+      .flatMap(new CreateMappingFromMarkedDuplicates<>())
+      .name("Create mapping to deduplicated vertices");
+    DataSet<V> deduplicatedVertices = annotatedVertices.filter(new RemoveMarkedDuplicates<>());
+    DataSet<V> otherVertices = graph.getVertices().filter(new Not<>(new ByLabel<>(label)));
+    DataSet<E> updatesEdges = graph.getEdges()
+      .leftOuterJoin(vertexToDedupVertex)
+      .where(new SourceId<>())
+      .equalTo(0)
+      .with(new UpdateEdgeSource<>())
+      .name("Update edges (source)")
+      .leftOuterJoin(vertexToDedupVertex)
+      .where(new TargetId<>())
+      .equalTo(0)
+      .with(new UpdateEdgeTarget<>())
```

Comment:
Annotating the update functions might enable flink to do both updates in parallel:
`@FunctionAnnotation.NonForwardedFields("targetId")`
Does this actually work with the interfaces? After all there is no "targetId" field...

Reply:
NonForwardedFields is not available for join functions, I added other annotations instead.